File: rust/src/account.rs
use hex;
use k256::{ecdsa::{self, signature::Signer}};

use crate::{
    transaction::{Transaction, SignedTransaction},
    block::{Block, SignedBlock},
    rpc::PB
};


pub struct Account {
    sk: ecdsa::SigningKey,
    nonce: u32,
}

impl Account {
    pub fn new(private_key: &str) -> Self {
        let priv_key_bytes = &hex::decode(private_key).unwrap();
        let sk = ecdsa::SigningKey::from_bytes(&priv_key_bytes).unwrap();
        Self {
            sk,
            nonce: 0
        }
    }

    pub fn get_address(&self) -> String {
        let vk = ecdsa::VerifyingKey::from(&self.sk);
        hex::encode(vk.to_bytes())
    }

    pub fn sign_transaction(&mut self, mut transaction: Transaction) -> SignedTransaction {
        transaction.set_nonce(self.nonce);
        self.nonce += 1;
        let signature: ecdsa::Signature = self.sk.sign(&hex::decode(transaction.hash()).unwrap());
        SignedTransaction::new(transaction, hex::encode(signature))
    }

    pub fn sign_block(&self, block: Block) -> SignedBlock {
        let signature: ecdsa::Signature = self.sk.sign(&hex::decode(block.hash()).unwrap());
        SignedBlock::new(block, hex::encode(signature))
    }
}

File: python/src/proto_chain/account.py
from functools import cached_property
from hashlib import sha256

from ecdsa import SigningKey, SECP256k1

from proto_chain.transaction import SignedTransaction
from proto_chain.block import SignedBlock


class Account:
    def __init__(self, private_key=None):
        if private_key is None:
            self._sk = SigningKey.generate(SECP256k1)
        else:
            self._sk = SigningKey.from_string(bytes.fromhex(private_key), SECP256k1)
        self._nonce = 0

    @cached_property
    def address(self):
        return self._sk.verifying_key.to_string("compressed").hex()

    def sign_transaction(self, transaction):
        transaction.set_nonce(self._nonce)
        self._nonce += 1
        return SignedTransaction(transaction, self._sk.sign_deterministic(bytes.fromhex(transaction.hash), hashfunc=sha256).hex())

    def sign_block(self, block):
        return SignedBlock(block, self._sk.sign_deterministic(bytes.fromhex(block.hash), hashfunc=sha256).hex())
